flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java

```diff
@@ -738,6 +738,25 @@ public ApplicationReport startAppMaster(
 			// add user code jars from the provided JobGraph
 			: jobGraph.getUserJars().stream().map(f -> f.toUri()).map(File::new).collect(Collectors.toSet());
 
+		// only for per job mode
+		if (jobGraph != null) {
+			final Collection<Tuple2<String, org.apache.flink.core.fs.Path>> userArtifacts = jobGraph.getUserArtifacts().entrySet().stream()
+				.map(entry -> Tuple2.of(entry.getKey(), new org.apache.flink.core.fs.Path(entry.getValue().filePath)))
+				.collect(Collectors.toList());
+
+			for (Tuple2<String, org.apache.flink.core.fs.Path> userArtifact : userArtifacts) {
```

Comment:
Can we use `for (Map.Entry<String, DistributedCache.DistributedCacheEntry> entry : jobGraph.getUserArtifacts().entrySet())` for the loop here? We may not need to create the userArtifacts list.

Reply:
@wangyang0918 
I updated the loop and added an integration test in yarn test module. Would you please take one more round of review?